File: bloom.py
# -*- coding: utf-8 -*-
'''
A simple counting bloom filter implementation using the MD5 hash function.
-- Jeethu Rao <jeethu@jeethurao.com>
'''
import array
import hashlib
import os
import itertools
import math

class BloomFilter( object ) :
    M = 16384
    K = 8
    HASH_ALGO = 'md5'
    def __init__( self, m=None, k=None, ivs=None, value=None ) :
        if m is not None :
            self.M = m
        if k is not None :
            self.K = k
        if value is not None :
            self._filter = array.array('B',value)
            l = len(self._filter)
            if l < self.M :
                self._filter.extend(itertools.repeat(0,self.M-l))
        else :
            self._filter = array.array('L',itertools.repeat(0,self.M))
        n_bits = int(math.floor(math.log(self.M,2)))
        self.n_bytes = int(math.ceil(n_bits/8.0))
        self.digest_size = hashlib.new(self.HASH_ALGO).digest_size
        if ivs is None :
            self.ivs = [os.urandom(16) for x in
                        range((self.n_bytes*self.K/self.digest_size)-1)]
            self.ivs.insert(0,'')
        else :
            self.ivs = ivs

    def hash( self, d ) :
        l = []
        h_l,l = [], []
        for iv in self.ivs :
            h = hashlib.new(self.HASH_ALGO,iv)
            h.update(d)
            h_l.append(h.digest())
        hashes = ''.join(h_l)
        pos = 0
        for x in range(self.K) :
            v = 0
            for n,x in enumerate(hashes[pos:pos+self.n_bytes]) :
                v += ord(x) << (n*8)
            l.append(v%self.M)
            pos += self.n_bytes
        return l

    def full( self ) :
        return sum(bool(x) for x in self._filter) == self.M

    def empty( self ) :
        return sum(bool(x) for x in self._filter) == 0

    def __nonzero__( self ) :
        return not self.empty()

    def add( self, value ) :
        hash = self.hash(value)
        if all(self._filter[x] for x in hash) :
            return
        for x in hash :
            self._filter[x] += 1
            if self._filter[x] == 0 :   # Prevent overflow
                self._filter[x] -= 1

    def _remove( self, value, hash, cnt=0 ) :
        for x in hash :
            if self._filter[x] :
                self._filter[x] -= 1
        # Check if the value is still present
        if cnt > 5 :
            raise RuntimeError, "Can't remove the element from this filter"
        if all(self._filter[x] for x in hash) :  
            return self._remove( value, hash, cnt+1 )
        return True

    def remove( self, value ) :
        hash = self.hash(value) 
        return self._remove( value, hash )
        
    def __contains__( self, value ) :
        return all(self._filter[x] for x in self.hash(value))

    def serialize( self ) :
        d = {
            'm'      : self.M,
            'k'      : self.K,
        }
        min_val = min(self._filter)
        cfilter = [x-min_val for x in self._filter.tolist()]
        if any(cfilter) :
            d['filter'] = cfilter
        if min_val :
            d['min_val'] = min_val
        if any(self.ivs) :
            d['ivs'] = self.ivs
        return d

    @classmethod
    def unserialize( cls, d ) :
        min_val = d.get('min_val',0)
        filter = [x+min_val for x in d.get('filter',[])]
        if not filter :
            filter = None
        ivs = d.get('ivs',[''])
        return cls( d['m'],d['k'],ivs,filter )

class SimpleBloomFilter(BloomFilter) :
    def __init__( self, capacity=100, err=0.1, ivs=None, value=None ) :
        '''
        @capacity: The capacity of the filter
        @err: The tolerable false positives rate ( 0 < err < 0.5 )
        '''
        m = int(math.ceil(capacity * math.log(err) / math.log(1.0 / 2**math.log(2))))
        k = int(math.floor(math.log(2) * m / capacity))
        BloomFilter.__init__( self, m, k, ivs, value )

File: js/bloom.js
/*
 * A Bloom filter in Javascript
 * 100% compatible with the python implementation.
 */
(function() {
    function unrle(s) {
        var el;
        var l = [];
        if(typeof s === "string") {
            return s;
        }
        for(var i=0,it=s.length;i<it;i++) {
            el = s[i];
            if(typeof el === "string") {
                l.push(el);
            } else {
                var r = [],c=el[0];
                for(var x=el[1];x>0;x--) {
                    r.push(c);
                }
                l.push(r.join(''));
            }
        }
        return l.join('');
    }

    var _HEX_MAP = function(){
        var l = '0123456789ABCDEF';
        var d = {};
        for(var i=0;i<l.length;i++) {
            d[l[i]] = i;
        }
        return d;
    }();

    function BloomFilter( d ) {
        if(d.hasOwnProperty('m')) {
            this.m = d.m;
        } else {
            this.m = 16384;
        }
        if(d.hasOwnProperty('k')) {
            this.k = d.k;
        } else {
            this.k = 8;
        }
        if(d.hasOwnProperty('ivs')) {
            this.ivs = d.ivs;
        } else {
            this.ivs = [''];
        }
        var filter = [];
        var i;
        if(d.hasOwnProperty('filter')&&d.filter) {
            var min_value = d.hasOwnProperty('min_val') ? d.min_val : 0;
            var data = unrle(d.filter);
            var data_len = data.length;
            if(data_len==this.m) {
                // 4 bits per value
                for(i=0;i<data_len;i++) {
                    filter.push(min_value + _HEX_MAP[data[i]]);
                }
            } else if(data_len==(this.m<<1)) {
                // 8 bits per value
                var val = 0;
                var state = 0;
                for(i=0;i<data_len;i++) {
                    if(state===0) {
                        val = _HEX_MAP[data[i]] << 4;
                    } else {
                        val += _HEX_MAP[data[i]];
                        filter.push(min_value + _HEX_MAP[data[i]]);
                    }
                    state = (state+1) & 1;
                }
            } else {
                throw "Error, incorrect filter len: " + data_len.toString();
            }
        } else {
            for(i=0;i<this.m;i++) {
                filter.push(0);
            }
        }
        this.filter = filter;
        this.calc_nbytes();
    }

    BloomFilter.prototype.calc_nbytes = function() {
        var n_bits = Math.floor(Math.LOG2E * Math.log(this.m));
        this.n_bytes = Math.ceil(n_bits/8.0);
    };

    // BloomFilter.prototype.sha256 = ???;

    BloomFilter.prototype.hexToInt = function( s ) {
        var v = 0;
        var a = s.split('');
        a.reverse();
        for(var i=0,j=s.length;i<j;i++) {
            v += _HEX_MAP[a[i]] << (i<<2);
        }
        return (v%this.m);
    };

    BloomFilter.prototype.hash = function( s ) {
        var l = [], h_l = [], i,j,hashes,v,pos=0;
        for(i=0,j=this.ivs.length;i<j;i++) {
            h_l.push(this.sha256(this.ivs[i] + s ));
        }
        hashes = h_l.join('').toUpperCase();
        for(i=0,j=this.k;i<j;i++) {
            l.push(this.hexToInt(hashes.slice(pos,pos+(this.n_bytes<<1))));
            pos += this.n_bytes<<1;
        }
        return l;
    };

    BloomFilter.prototype.hasKey = function( s ) {
        var cnt=0;
        var h = this.hash(s);
        for(var i=0,j=h.length;i<j;i++) {
            if(!this.filter[h[i]]) {
                return false;
            }
        }
        return true;
    };

    BloomFilter.prototype.add = function( value ) {
        var hash = this.hash(value), filter=this.filter,i=0,j=hash.length,cnt=0;
        for(;i<j;i++) {
            if(filter[hash[i]]) {
                cnt += 1;
            } else {
                break;
            }
        }
        if(cnt==j) {
            return;
        }
        for(i=0;i<j;i++) {
            filter[hash[i]] += 1;
        }
    };

    BloomFilter.prototype._remove = function( hash, cnt ) {
        var i=0,j = hash.length, filter=this.filter;
        for(;i<j;i++) {
            if(filter[hash[i]]) {
                filter[hash[i]] -= 1;
            }
        }
        if(cnt>5) {
            throw "Can't remove element from this filter";
        }
        for(i=0;i<j;i++) {
            if(!filter[hash[i]]) {
                return true;
            }
        }
        return this._remove( hash, cnt+1 );
    };

    BloomFilter.prototype.remove = function( value ) {
        var hash = this.hash(value);
        return this._remove( hash, 0 );
    };

    function _filter_sum( that ) {
        var s=0,filter=that.filter;
        for(var i=0,j=that.m;i<j;i++) {
            if(filter[i]) {
                s+=1;
            }
        }
        return s;
    };

    BloomFilter.prototype.empty = function() {
        return _filter_sum(this)==0;
    };

    BloomFilter.prototype.full = function() {
        return _filter_sum(this)==this.m;
    };

    BloomFilter.unrle = unrle;

    window.BloomFilter = BloomFilter;
})();
